### PR TITLE
[wgsl] Initial commit of WGSL spec

### PR DESCRIPTION
**Pull request by dj2**
This is the start of the WebGPU Shading Language specification.

### CODE REVIEW

**Comment by Kangz** [comment]
Didn't we agree to make a different repo at the F2F? [notes](https://docs.google.com/document/d/1vQPA1JSOvfCHjBrkAEDLA1qCqQXe72vGen_1quoHZV8/edit#bookmark=id.drzbv8ywzkpm)

**Comment by dj2** [comment]
This was discussed at the meeting on Monday and it was decided to put it in the main repo. [link](https://docs.google.com/document/d/1edJjmTyqZuCzVxh0J8TT0HVkXenPPJysAsizk785I2M/edit#bookmark=id.lpbs4fz09oky)

**Comment by Kangz** [comment]
Ok, some D3D folks were interested in following just the WGSL stuff but I don't think you can watch only specific tags on GitHub.

**Comment by dj2** [comment]
The plan is to have the `[wgsl]` text in the pull request titles. So, in theory you can filter on `[gpuweb/gpuweb] [wgsl]` in the subject

File: wgsl/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+The specification is written using [Bikeshed](https://tabatkins.github.io/bikeshed).
+
+If you have bikeshed installed locally, you can generate the specification with:
+
+```
+make
+```
+
+This simply runs bikeshed on the `index.bs` file.
```
Comment (dneto0) [inline wgsl/README.md]: Nitpick:   Avoid "simply".  It's not simple the first time you do it.  Installing bikeshed might be part of that first time.

**Comment by kvark** [comment]
@Kangz that's actually a point that was not considered during the call. It comes down to the fact that, even though there are changes that would need to be done in both the spec and the language, there is a lot of work on the language itself that doesn't in any way involve the API spec.

**Comment by dj2** [comment]
It would be nice if the CTS was also in this repo and then committing WGSL changes would have their tests in one commit.

File: wgsl/README.md (hunk 1 above)
Comment (litherum) [reply wgsl/README.md]: Don't need to install Bikeshed. https://api.csswg.org/bikeshed/

File: wgsl/.pr-preview.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+    "src_file": "index.bs",
+    "type": "bikeshed",
+    "params": {
+        "force": 1
```
Comment (litherum) [inline wgsl/.pr-preview.json]: I don't think we want this. Errors are bad.

**Comment by litherum** [comment]
Beware, the draft spec is large enough that GitHub doesn't preview it by default. You have to click to see the draft.

File: wgsl/.pr-preview.json (hunk 2 above)
Comment (dj2) [reply wgsl/.pr-preview.json]: I thought this file was just for preview of the PR? In which case, don't we want to force it so we can see the rendering?

File: wgsl/README.md (hunk 1 above)
Comment (dj2) [reply wgsl/README.md]: Installing bikeshed is optional. You can use the `make online` variant if you don't want to install bikeshed.

**Comment by Kangz** [comment]
+1 for merging and changing things later. @beaufortfrancois what do we need to do to set up an automatically rendered version like for the WebGPU spec?

**Comment by lain-dono** [comment]
Does the adoption of WGSL as part of the standard mean a complete rejection of SPIR-V?

**Comment by dj2** [comment]
Not sure what you mean by complete rejection? Each construct in WGSL is defined in terms of SPIR-V (or will be for the ones that I haven't gotten to yet)

**Comment by dneto0** [comment]
Personally I'd prefer in a different repo.  I agree 100% with @kvark about the reason.
But I don't want to block progress at this point.

**Pull request merged**